### PR TITLE
zfs-unstable: 2.0.4 -> 2.1.0-rc1

### PR DESCRIPTION
**Pull request by ali-abrar**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
I am not sure if we want to update to a rc yet.

**Comment by hmenke** [comment]
According to the release notes:

> Linux: compatible with 3.10 - 5.11 kernels

Why did you bump it to 5.12?

**Comment by hmenke** [comment]
@SuperSandro2000 I think it's okay to update `zfsUnstable` to a rc. That's what it is for.

**Comment by ali-abrar** [comment]
> According to the release notes:
> 
> > Linux: compatible with 3.10 - 5.11 kernels
> 
> Why did you bump it to 5.12?

This version includes some [fixes for 5.12](https://github.com/openzfs/zfs/pull/11639).

**Comment by hmenke** [comment]
> This version includes some [fixes for 5.12](https://github.com/openzfs/zfs/pull/11639) and can be used with that kernel (though at the users own risk, as neither the kernel nor the openzfs release candidate are final yet).

I'm strongly against this.  We should at least uphold the upstream guarantees regarding supported kernel versions. While I don't think there have been reports of data loss, there were definitely some serious issues with unsupported kernels in the past, see e.g. @Atemu's report here https://github.com/danielfullmer/robotnix/issues/74

File: pkgs/os-specific/linux/zfs/default.nix [hunk 1]
```diff
@@ -207,12 +207,12 @@ in {
 
   zfsUnstable = common {
     # check the release notes for compatible kernels
-    kernelCompatible = kernel.kernelAtLeast "3.10" && kernel.kernelOlder "5.12";
+    kernelCompatible = kernel.kernelAtLeast "3.10" && kernel.kernelOlder "5.13";
```
Comment (hmenke) [inline pkgs/os-specific/linux/zfs/default.nix:210]:
```suggestion
    kernelCompatible = kernel.kernelAtLeast "3.10" && kernel.kernelOlder "5.12";
```

**Review by hmenke — CHANGES_REQUESTED** [review]

**Comment by hmenke** [comment]
Result of `nixpkgs-review pr 118616` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages marked as broken and skipped:</summary>
  <ul>
    <li>linuxPackages-libre.zfsUnstable</li>
    <li>linuxPackages_hardkernel_4_14.zfsUnstable</li>
    <li>linuxPackages_hardkernel_latest.zfsUnstable</li>
    <li>linuxPackages_latest-libre.zfsUnstable</li>
  </ul>
</details>
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>linuxPackages_lqx.zfsUnstable</li>
    <li>linuxPackages_zen.zfsUnstable</li>
    <li>zfsUnstable</li>
  </ul>
</details>
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>linuxPackages.zfsUnstable (linuxPackages_5_10.zfsUnstable)</li>
    <li>linuxPackages_4_14.zfsUnstable</li>
    <li>linuxPackages_4_19.zfsUnstable</li>
    <li>linuxPackages_4_4.zfsUnstable</li>
    <li>linuxPackages_4_9.zfsUnstable</li>
    <li>linuxPackages_5_11.zfsUnstable (linuxPackages_latest.zfsUnstable)</li>
    <li>linuxPackages_5_4.zfsUnstable</li>
    <li>linuxPackages_hardened.zfsUnstable</li>
    <li>linuxPackages_latest_hardened.zfsUnstable</li>
    <li>linuxPackages_latest_xen_dom0.zfsUnstable</li>
    <li>linuxPackages_latest_xen_dom0_hardened.zfsUnstable</li>
    <li>linuxPackages_testing_bcachefs.zfsUnstable</li>
    <li>linuxPackages_xen_dom0.zfsUnstable</li>
    <li>linuxPackages_xen_dom0_hardened.zfsUnstable</li>
  </ul>
</details>

**Comment by Mic92** [comment]
Does `zfsUnstable` really fails to build @hmenke ?

**Comment by hmenke** [comment]
@Mic92 Yep, it fails:
```
error: cycle detected in build of '/nix/store/bbz7s8ps672amx1rxw7p53a5yw84a1d6-zfs-user-2.1.0-rc1.drv' in the references of output 'lib' from output 'out' 
```

**Comment by Luflosi** [comment]
Same problem with rc2.

**Comment by Mic92** [comment]
https://github.com/openzfs/zfs/releases/tag/zfs-2.1.0-rc2

**Comment by hmenke** [comment]
I found the offender: https://github.com/openzfs/zfs/blob/zfs-2.1.0-rc3/lib/libzfs/libzfs_pool.c#L4806-L4807

ZFS has introduced an new compatibility property (https://github.com/openzfs/zfs/commit/658fb8020f0501435516baeea7004575d640649b) and some predefined feature sets are stored in
1) ZPOOL_SYSCONF_COMPAT_D (eg: /etc/zfs/compatibility.d) or
2) ZPOOL_DATA_COMPAT_D (eg: /usr/share/zfs/compatibility.d).
Both of them are rewritten to `$out/...` in the Nix build.

The problem is that both the `zpool` command in `$out` and `libzfs` in `$lib` use these directories, so it is not possible to put them in either without creating a cycle. To me it looks like we have to give up the `out`/`lib` split.

<details>
<summary><code>grep -r 'COMPAT_D\b'</code></summary>

```
cmd/zpool/zpool_main.c:             "appropriate file in " ZPOOL_SYSCONF_COMPAT_D " or "
cmd/zpool/zpool_main.c:             ZPOOL_DATA_COMPAT_D ".\n"));
cmd/zpool/zpool_main.c:             ZPOOL_SYSCONF_COMPAT_D " or " ZPOOL_DATA_COMPAT_D ".\n"));
lib/libzfs/libzfs_pool.c: *   1) ZPOOL_SYSCONF_COMPAT_D (eg: /etc/zfs/compatibility.d) or
lib/libzfs/libzfs_pool.c: *   2) ZPOOL_DATA_COMPAT_D (eg: /usr/share/zfs/compatibility.d).
lib/libzfs/libzfs_pool.c:       sdirfd = open(ZPOOL_SYSCONF_COMPAT_D, O_DIRECTORY | O_PATH | O_CLOEXEC);
lib/libzfs/libzfs_pool.c:       ddirfd = open(ZPOOL_DATA_COMPAT_D, O_DIRECTORY | O_PATH | O_CLOEXEC);
include/sys/fs/zfs.h:#define    ZPOOL_SYSCONF_COMPAT_D  SYSCONFDIR "/zfs/compatibility.d"
include/sys/fs/zfs.h:#define    ZPOOL_DATA_COMPAT_D     PKGDATADIR "/compatibility.d"
```

</details>

**Comment by hmenke** [comment]
@Mic92 @jonringer What's your opinion on my previous comment? There are no derivations in nixpkgs that reference `zfs.lib`  (or `zfs.dev` for that matter), so merging `$out` and `$lib` should not be a problem and is a simple fix.

**Comment by jonringer** [comment]
I'm okay with merging them. Just please add a comment that having a `dev` and `out` output isn't possible without cycles.

**Pull request closed without merging**